codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -285,7 +301,22 @@ public ChannelFuture writePushPromise(ChannelHandlerContext ctx, int streamId, i
             // Writing headers may fail during the encode state if they violate HPACK limits.
             Throwable failureCause = future.cause();
             if (failureCause == null) {
+                // This is best-afford as the operation is async it may still fail later on. We just update it here
+                // directly as otherwise it may be challenging for the end-user to co-ordinate writes that depend
+                // on the state.
                 stream.pushPromiseSent();
+
+                if (!future.isDone()) {
+                    future.addListener(new ChannelFutureListener() {
```

Comment:
This entire `if` block could be extracted into a method for reduced duplication?

Reply:
sure let me do

Reply:
done